### PR TITLE
cloud: allow connect to cloud with coaps scheme

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: api/cloud/oc_cloud_apis.c [hunk 1]
```diff
@@ -698,6 +699,26 @@ cloud_access_refresh_access_token(oc_endpoint_t *endpoint, const char *uid,
...
 }
+
+bool
+cloud_send_ping(oc_endpoint_t *endpoint, uint16_t timeout_seconds,
+                oc_response_handler_t handler, void *user_data)
+{
+  if (!oc_tls_connected(endpoint)) {
+    return false;
+  }
+#ifdef OC_TCP
```
Comment (Danielius1922) [inline api/cloud/oc_cloud_apis.c:710]:
```C
#ifdef OC_TCP
  if (endpoint->flags & TCP) {
    return oc_send_ping(false, endpoint, timeout_seconds, handler, user_data);
  }
#endif /* OC_TCP */
 
  return oc_do_get_with_timeout(OC_RSRVD_RD_URI, endpoint, NULL,
                                  timeout_seconds, handler, LOW_QOS, user_data);
}
```

File: api/cloud/oc_cloud.c [hunk 2]
```diff
@@ -327,8 +327,11 @@ oc_cloud_manager_restart(oc_cloud_context_t *ctx)
   OC_DBG("[CM] oc_cloud_manager_restart\n");
 #ifdef OC_SESSION_EVENTS
   if (ctx->cloud_ep_state == OC_SESSION_CONNECTED) {
+    bool is_tcp = (ctx->cloud_ep->flags & TCP) != 0;
     cloud_close_endpoint(ctx->cloud_ep);
-    return;
+    if (is_tcp) {
```
Comment (Danielius1922) [inline api/cloud/oc_cloud.c:332]: There is an non-tcp connection that doesn't trigger `cloud_ep_session_event_handler`?

File: api/cloud/oc_cloud_internal.h [hunk 3]
```diff
@@ -278,6 +278,19 @@ bool cloud_set_retry(const uint8_t retry_timeout[], size_t retry_timeout_size);
...
 size_t cloud_get_retry(uint8_t *buffer, size_t buffer_size);
 
+/**
+ * @brief Send a ping over the cloud connected connection
+ *
+ * @param endpoint endpoint to be used
+ * @param timeout_seconds timeout for the ping
+ * @param handler the response handler
+ * @param user_data the user data to be conveyed to the response handler
+ * @return true
```
Comment (Danielius1922) [inline api/cloud/oc_cloud_internal.h:288]: I guess true on success and false on failure?

File: api/cloud/oc_cloud_apis.c [hunk 4]
```diff
@@ -698,6 +699,26 @@ cloud_access_refresh_access_token(oc_endpoint_t *endpoint, const char *uid,
 
   return oc_do_post();
 }
+
+bool
+cloud_send_ping(oc_endpoint_t *endpoint, uint16_t timeout_seconds,
+                oc_response_handler_t handler, void *user_data)
+{
+  if (!oc_tls_connected(endpoint)) {
```
Comment (Danielius1922) [inline api/cloud/oc_cloud_apis.c:707]: Every other call to `oc_tls_connected` seems to be under OC_SECURITY macro. Isn't it missing here?

File: api/cloud/oc_cloud.c (hunk 2 above)
Comment (jkralik) [reply api/cloud/oc_cloud.c:332]: DTLS and UDP connectionn doesn't trigger `cloud_ep_session_event_handler`

File: api/cloud/oc_cloud_apis.c (hunk 1 above)
Comment (jkralik) [reply api/cloud/oc_cloud_apis.c:729]: fixed

File: api/cloud/oc_cloud_internal.h (hunk 3 above)
Comment (jkralik) [reply api/cloud/oc_cloud_internal.h]: fixed

File: api/cloud/oc_cloud_apis.c (hunk 4 above)
Comment (jkralik) [reply api/cloud/oc_cloud_apis.c]: fixed

**Review by Danielius1922 — APPROVED** [review]

**Pull request merged**